pkg/apis/networking/v1alpha1/clusteringress_types.go

```diff
@@ -0,0 +1,376 @@
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the ClusterIngress.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec ClusterIngressSpec `json:"spec,omitempty"`
+
+	// Status is the current state of the ClusterIngress.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status ClusterIngressStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterIngressList is a collection of ClusterIngress.
+type ClusterIngressList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of ClusterIngress.
+	Items []ClusterIngress `json:"items"`
+}
+
+// ClusterIngressSpec describes the ClusterIngress the user wishes to exist.
+type ClusterIngressSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// TLS configuration. Currently the ClusterIngress only supports a single TLS
+	// port, 443. If multiple members of this list specify different hosts, they
+	// will be multiplexed on the same port according to the hostname specified
+	// through the SNI TLS extension, if the ingress controller fulfilling the
+	// ingress supports SNI.
+	// +optional
+	TLS []ClusterIngressTLS `json:"tls,omitempty"`
+
+	// A list of host rules used to configure the ClusterIngress.
+	// +optional
+	Rules []ClusterIngressRule `json:"rules,omitempty"`
+}
+
+// ClusterIngressTLS describes the transport layer security associated with an ClusterIngress.
+type ClusterIngressTLS struct {
+	// Hosts are a list of hosts included in the TLS certificate. The values in
+	// this list must match the name/s used in the tlsSecret. Defaults to the
+	// wildcard host setting for the loadbalancer controller fulfilling this
+	// ClusterIngress, if left unspecified.
+	// +optional
+	Hosts []string `json:"hosts,omitempty"`
+	// SecretName is the name of the secret used to terminate SSL traffic.
+	// Field is left optional to allow SSL routing based on SNI hostname alone.
+	// +optional
+	SecretName string `json:"secretName,omitempty"`
+
+	// SecretNamespace is the namespace of the secret used to terminate SSL traffic.
+	// Field is left optional to allow SSL routing based on SNI hostname alone.
```

Comment:
I don't understand this comment. Don't we need the TLS certificate for a ClusterIngressTLS?